File: Java/src/DP.java
import java.util.*;

public class DP {

    public static int knapsack01(int[] w, int[] v, int W) {
        int[] dp = new int[W+1];
        for (int i=0;i<w.length;i++) {
            for (int cap=W; cap>=w[i]; cap--) {
                dp[cap] = Math.max(dp[cap], dp[cap-w[i]] + v[i]);
            }
        }
        return dp[W];
    }

    public static int lisLength(int[] arr) {
        ArrayList<Integer> t = new ArrayList<>();
        for (int x : arr) {
            int pos = Collections.binarySearch(t, x);
            if (pos < 0) pos = -pos - 1;
            if (pos == t.size()) t.add(x);
            else t.set(pos, x);
        }
        return t.size();
    }
}

File: Python/src/dp.py
# Dynamic programming: 0/1 knapsack and LIS
from typing import List

def knapsack_01(weights: List[int], values: List[int], W: int) -> int:
    n = len(weights)
    dp = [0]*(W+1)
    for i in range(n):
        w = weights[i]; v = values[i]
        for cap in range(W, w-1, -1):
            dp[cap] = max(dp[cap], dp[cap-w] + v)
    return dp[W]

def lis_length(arr: List[int]) -> int:
    # O(n log n) patience algorithm
    import bisect
    tails = []
    for x in arr:
        i = bisect.bisect_left(tails, x)
        if i == len(tails):
            tails.append(x)
        else:
            tails[i] = x
    return len(tails)
